File: js/promptBuilder.js
import { grammarTemplates, vocabularyTemplates, difficultyLevels, baseRole } from './templates.js';
import { ankiDataManager } from './ankiData.js';

export class SpanishTutorPromptBuilder {
    constructor(userPreferences) {
        this.preferences = userPreferences;
    }

    build() {
        let prompt = baseRole + "\n\n";

        // 1. Difficulty Level (Required)
        const level = this.preferences.difficultyLevel || 'B1';
        const levelDescription = difficultyLevels[level] || difficultyLevels['B1'];

        prompt += `CURRENT DIFFICULTY LEVEL: ${level}\n`;
        prompt += `${levelDescription}\n\n`;

        // 2. Anki Spaced Repetition Vocabulary Guidance (If available)
        const ankiGuidance = ankiDataManager.getVocabularyGuidance();
        if (ankiGuidance.totalWords > 0) {
            prompt += "ANKI VOCABULARY SCAFFOLDING:\n";
            prompt += "The user has imported their Anki deck data. Use this to create an optimal learning experience:\n\n";

            // Mastered words (use freely as scaffolding)
            if (ankiGuidance.mastered.length > 0) {
                const masteredSample = this.sampleWords(ankiGuidance.mastered, 30);
                prompt += `MASTERED WORDS (use freely - these are well-known to the user):\n`;
                prompt += `${masteredSample.join(', ')}\n`;
                if (ankiGuidance.mastered.length > 30) {
                    prompt += `...and ${ankiGuidance.mastered.length - 30} more mastered words\n`;
                }
                prompt += "\n";
            }

            // Familiar words (good for scaffolding)
            if (ankiGuidance.familiar.length > 0) {
                const familiarSample = this.sampleWords(ankiGuidance.familiar, 30);
                prompt += `FAMILIAR WORDS (comfortable for the user - use to support learning):\n`;
                prompt += `${familiarSample.join(', ')}\n`;
                if (ankiGuidance.familiar.length > 30) {
                    prompt += `...and ${ankiGuidance.familiar.length - 30} more familiar words\n`;
                }
                prompt += "\n";
            }

            // Learning words (current focus - use with support)
            if (ankiGuidance.learning.length > 0) {
                const learningSample = this.sampleWords(ankiGuidance.learning, 20);
                prompt += `CURRENTLY LEARNING (use these more frequently to reinforce):\n`;
                prompt += `${learningSample.join(', ')}\n`;
                if (ankiGuidance.learning.length > 20) {
                    prompt += `...and ${ankiGuidance.learning.length - 20} more learning words\n`;
                }
                prompt += "\n";
            }

            // New/struggling words (introduce carefully with context)
            if (ankiGuidance.new.length > 0) {
                const newSample = this.sampleWords(ankiGuidance.new, 15);
                prompt += `NEW/CHALLENGING WORDS (introduce carefully with familiar word support):\n`;
                prompt += `${newSample.join(', ')}\n`;
                if (ankiGuidance.new.length > 15) {
                    prompt += `...and ${ankiGuidance.new.length - 15} more new words\n`;
                }
                prompt += "\n";
            }

            prompt += "SCAFFOLDING STRATEGY:\n";
            prompt += "- Build sentences using MASTERED and FAMILIAR words as the foundation\n";
            prompt += "- Naturally incorporate LEARNING words to reinforce retention\n";
            prompt += "- Introduce NEW words in context with lots of familiar vocabulary support\n";
            prompt += "- Create connections between new words and words the user already knows well\n";
            prompt += "- This approach maximizes comprehension while gently expanding vocabulary\n\n";
        }

        // 3. Target Vocabulary (Optional)
        if (this.preferences.targetVocabulary) {
            prompt += "TARGET VOCABULARY:\n";
            prompt += `Try to naturally include the following words/phrases in your responses: ${this.preferences.targetVocabulary}\n\n`;
        }

        // 4. Grammar Focus (Optional)
        if (this.preferences.selectedGrammar && this.preferences.selectedGrammar.length > 0) {
            prompt += "GRAMMAR FOCUS:\n";
            prompt += "Prioritize using or eliciting the following grammar concepts:\n";
            this.preferences.selectedGrammar.forEach(g => prompt += `- ${g}\n`);
            prompt += "\n";
        }

        // 5. Custom Instructions / Scenarios (Optional - High Priority)
        if (this.preferences.customInstructions) {
            prompt += "USER CUSTOM REQUEST (PRIORITY):\n";
            prompt += `The user specifically asked: "${this.preferences.customInstructions}".\nADAPT YOUR CONVERSATION TO FULFILL THIS REQUEST ABOVE ALL ELSE.\n`;
        }

        // General Rules
        prompt += "IMPORTANT RULES:\n";
        prompt += "1. Speak ONLY in Spanish. No English unless requested.\n";
        prompt += "2. Keep responses concise (1-3 sentences) to encourage a back-and-forth dialogue.\n";
        prompt += "3. REFLECTIVE FEEDBACK: If the user makes a mistake, rephrase it correctly in your response naturally, or ask a clarifying question. Do not lecture.\n";

        return prompt;
    }

    /**
     * Helper method to sample random words from a list
     * @param {Array} words - Array of words to sample from
     * @param {number} count - Number of words to sample
     * @returns {Array} Sampled words
     */
    sampleWords(words, count) {
        if (words.length <= count) {
            return words;
        }

        // Shuffle and take first 'count' items
        const shuffled = [...words].sort(() => Math.random() - 0.5);
        return shuffled.slice(0, count);
    }

    static parseNaturalInstruction(instruction) {
        const result = {
            grammarFocus: [],
            vocabularyFocus: [],
            customInstructions: instruction // Always keep the raw instruction
        };

        if (!instruction) return result;

        const lower = instruction.toLowerCase();

        // Simple keyword matching
        // Grammar
        if (lower.includes('subjunctive')) result.grammarFocus.push('subjunctive');
        if (lower.includes('command')) result.grammarFocus.push('subjunctiveCommands');
        if (lower.includes('past') || lower.includes('preterite') || lower.includes('imperfect')) result.grammarFocus.push('pastTense');
        if (lower.includes('future')) result.grammarFocus.push('future');
        if (lower.includes('conditional') || lower.includes('would')) result.grammarFocus.push('conditional');
        if (lower.includes('por') && lower.includes('para')) result.grammarFocus.push('porVsPara');
        if (lower.includes('ser') && lower.includes('estar')) result.grammarFocus.push('serVsEstar');

        // Vocabulary
        if (lower.includes('restaurant') || lower.includes('food') || lower.includes('eat')) result.vocabularyFocus.push('restaurant');
        if (lower.includes('travel') || lower.includes('trip') || lower.includes('hotel')) result.vocabularyFocus.push('travel');
        if (lower.includes('shop') || lower.includes('buy')) result.vocabularyFocus.push('shopping');
        if (lower.includes('doctor') || lower.includes('sick') || lower.includes('medical')) result.vocabularyFocus.push('medical');
        if (lower.includes('business') || lower.includes('work') || lower.includes('job')) result.vocabularyFocus.push('business');
        if (lower.includes('family') || lower.includes('parent') || lower.includes('brother')) result.vocabularyFocus.push('family');
        if (lower.includes('hobby') || lower.includes('sport') || lower.includes('music')) result.vocabularyFocus.push('hobbies');

        return result;
    }
}

File: SolturaIOS/Soltura/Managers/PromptBuilder.swift
//
//  PromptBuilder.swift
//  Soltura
//
//  Builds system prompts for Spanish conversation AI
//

import Foundation

class PromptBuilder {
    private let preferences: UserPreferences

    init(preferences: UserPreferences) {
        self.preferences = preferences
    }

    func build() -> String {
        var prompt = Templates.baseRole + "\n\n"

        // 1. Difficulty Level
        let level = preferences.proficiencyLevel
        let levelDescription = Templates.difficultyLevels[level.rawValue] ?? Templates.difficultyLevels["Auto"]!

        prompt += "CURRENT DIFFICULTY LEVEL: \(level.rawValue)\n"
        prompt += "\(levelDescription)\n\n"

        // 2. Anki Vocabulary Scaffolding
        if !preferences.ankiVocabulary.isEmpty {
            let stats = preferences.getAnkiStats()
            prompt += buildAnkiSection(stats: stats)
        }

        // 3. Target Vocabulary
        if !preferences.targetVocabulary.isEmpty {
            prompt += "TARGET VOCABULARY:\n"
            prompt += "Try to naturally include the following words/phrases in your responses: \(preferences.targetVocabulary.joined(separator: ", "))\n\n"
        }

        // 4. Grammar Focus
        if !preferences.grammarFocus.isEmpty {
            prompt += "GRAMMAR FOCUS:\n"
            prompt += "Prioritize using or eliciting the following grammar concepts:\n"
            for grammar in preferences.grammarFocus {
                prompt += "- \(grammar.displayName)\n"
            }
            prompt += "\n"
        }

        // 5. Scenario or Custom Goal
        if !preferences.selectedScenario.isEmpty {
            if let scenario = ScenarioData.getScenario(byId: preferences.selectedScenario) {
                prompt += "CONVERSATION SCENARIO (HIGH PRIORITY):\n"
                prompt += "\(scenario.instruction)\n\n"
            }
        } else if !preferences.customGoal.isEmpty {
            prompt += "USER CUSTOM REQUEST (PRIORITY):\n"
            prompt += "The user specifically asked: \"\(preferences.customGoal)\".\n"
            prompt += "ADAPT YOUR CONVERSATION TO FULFILL THIS REQUEST ABOVE ALL ELSE.\n\n"
        }

        // General Rules
        prompt += "IMPORTANT RULES:\n"
        prompt += "1. Speak ONLY in Spanish. No English unless requested.\n"
        prompt += "2. Keep responses concise (1-3 sentences) to encourage a back-and-forth dialogue.\n"
        prompt += "3. REFLECTIVE FEEDBACK: If the user makes a mistake, rephrase it correctly in your response naturally, or ask a clarifying question. Do not lecture.\n"

        return prompt
    }

    private func buildAnkiSection(stats: (mastered: Int, familiar: Int, learning: Int, new: Int)) -> String {
        var section = "ANKI VOCABULARY SCAFFOLDING:\n"
        section += "The user has imported their Anki deck data. Use this to create an optimal learning experience:\n\n"

        let mastered = preferences.ankiVocabulary.filter { $0.masteryLevel == .mastered }
        let familiar = preferences.ankiVocabulary.filter { $0.masteryLevel == .familiar }
        let learning = preferences.ankiVocabulary.filter { $0.masteryLevel == .learning }
        let new = preferences.ankiVocabulary.filter { $0.masteryLevel == .new }

        // Mastered words
        if !mastered.isEmpty {
            let sample = sampleWords(from: mastered, count: 15)
            section += "MASTERED WORDS (use freely - these are well-known to the user):\n"
            section += "\(sample.map { $0.front }.joined(separator: ", "))\n"
            if mastered.count > 15 {
                section += "...and \(mastered.count - 15) more mastered words\n"
            }
            section += "\n"
        }

        // Familiar words
        if !familiar.isEmpty {
            let sample = sampleWords(from: familiar, count: 15)
            section += "FAMILIAR WORDS (comfortable for the user - use to support learning):\n"
            section += "\(sample.map { $0.front }.joined(separator: ", "))\n"
            if familiar.count > 15 {
                section += "...and \(familiar.count - 15) more familiar words\n"
            }
            section += "\n"
        }

        // Learning words
        if !learning.isEmpty {
            let sample = sampleWords(from: learning, count: 12)
            section += "CURRENTLY LEARNING (use these more frequently to reinforce):\n"
            section += "\(sample.map { $0.front }.joined(separator: ", "))\n"
            if learning.count > 12 {
                section += "...and \(learning.count - 12) more learning words\n"
            }
            section += "\n"
        }

        // New words
        if !new.isEmpty {
            let sample = sampleWords(from: new, count: 10)
            section += "NEW/CHALLENGING WORDS (introduce carefully with familiar word support):\n"
            section += "\(sample.map { $0.front }.joined(separator: ", "))\n"
            if new.count > 10 {
                section += "...and \(new.count - 10) more new words\n"
            }
            section += "\n"
        }

        section += "SCAFFOLDING STRATEGY:\n"
        section += "- Build sentences using MASTERED and FAMILIAR words as the foundation\n"
        section += "- Naturally incorporate LEARNING words to reinforce retention\n"
        section += "- Introduce NEW words in context with lots of familiar vocabulary support\n"
        section += "- Create connections between new words and words the user already knows well\n"
        section += "- This approach maximizes comprehension while gently expanding vocabulary\n\n"

        return section
    }

    private func sampleWords(from cards: [AnkiCard], count: Int) -> [AnkiCard] {
        if cards.count <= count {
            return cards
        }

        var result = Array(cards.prefix(count))
        for i in count..<cards.count {
            let j = Int.random(in: 0...i)
            if j < count {
                result[j] = cards[i]
            }
        }
        return result
    }
}

// MARK: - Templates

struct Templates {
    static let baseRole = "You are a friendly, patient, and encouraging native Spanish tutor. Your goal is to help the user practice conversational Spanish. You should maintain a natural conversation flow while subtly guiding them to practice specific topics if requested. Never lecture. Always keep the conversation going with relevant follow-up questions."

    static let difficultyLevels: [String: String] = [
        "Auto": """
**ADAPTIVE DIFFICULTY MODE - Dynamic Level Matching:**

Your primary task is to ANALYZE and MIRROR the user's demonstrated Spanish proficiency in real-time. This is a dynamic, conversation-by-conversation adaptation system.

**Analysis Framework - Evaluate Each User Message For:**

1. **Grammar Complexity:**
   - Which verb tenses are they using? (present only, past tenses, subjunctive, conditionals?)
   - Sentence structure complexity (simple SVO, compound, subordinate clauses?)
   - Error patterns (consistent vs random, which structures cause problems?)

2. **Vocabulary Sophistication:**
   - Basic vs advanced word choices?
   - Topic-specific vocabulary usage?
   - Reliance on cognates vs native Spanish words?

3. **Fluency Indicators:**
   - Message length and elaboration
   - Confidence in expression (hesitations, corrections, code-switching?)
   - Ability to express nuanced ideas

**Adaptive Response Strategy:**

- **MIRROR their demonstrated level** in your responses
- If they use present tense only → respond with present tense, simple structures
- If they use past tenses → incorporate similar past tense complexity
- If they attempt subjunctive → validate with subjunctive use, but keep it accessible
- If they write short, simple sentences → match with similar brevity
- If they elaborate with complex ideas → respond with equivalent complexity

**Dynamic Adjustment Rules:**

- Start neutral (B1-level) in your first response, then adapt immediately
- Each message is a new data point - continuously recalibrate
- Stay within ±1 level of their demonstrated proficiency
- Introduce slightly more advanced structures occasionally (10-20% above their level) to encourage growth, but not overwhelm
- If user makes errors, gently model correct forms without explicit correction
- Adapt vocabulary density to match theirs

**Key Principle:** You are a conversational mirror that matches their energy, complexity, and proficiency while subtly guiding them upward. Think "meet them where they are, then gently pull them forward."
""",

        "A1": """
**LLM Output Guidelines:**
- Use only present tense (regular and high-frequency irregulars: ser, estar, tener, ir)
- Keep sentences simple: subject-verb-object order
- Use basic vocabulary with clear noun-adjective agreement
- Employ simple questions with common question words (qué, quién, dónde, cuándo)
- Stick to concrete, present-time topics
- Avoid compound sentences; use mainly "y" and "pero" for connecting ideas
- Use reflexive verbs only for daily routines (levantarse, ducharse)
""",

        "A2": """
**LLM Output Guidelines:**
- Introduce past tense (pretérito and imperfecto) and simple future (ir a + infinitive)
- Use direct and indirect object pronouns in straightforward contexts
- Employ comparatives (más/menos que) and basic superlatives
- Include reflexive verbs more broadly
- Create slightly longer sentences with porque, cuando, después
- Discuss familiar topics in present, past, and near future
- Keep subjunctive minimal or absent
""",

        "B1": """
**LLM Output Guidelines:**
- Mix present, preterite, imperfect, and present perfect naturally
- Introduce present subjunctive in common contexts (espero que, es importante que)
- Use commands (imperative) appropriately
- Combine direct/indirect object pronouns (se lo, se la)
- Employ relative pronouns (que, quien) and time connectors (aunque, mientras, antes de)
- Discuss abstract topics with cause-effect relationships
- Begin conditional sentences (Si estudias, aprobarás)
""",

        "B2": """
**LLM Output Guidelines:**
- Use all past tenses (preterite, imperfect, present perfect, pluperfect) with nuance
- Employ imperfect subjunctive in hypothetical clauses (Si tuviera…, me gustaría que…)
- Use conditional tenses (simple and perfect) for speculation and politeness
- Include passive voice (ser + past participle) and impersonal "se"
- Create complex sentences with subordinate clauses
- Use advanced connectors (sin embargo, por lo tanto, a pesar de que)
- Express subtle differences in mood and register
""",

        "C1": """
**LLM Output Guidelines:**
- Master all subjunctive forms (present, imperfect, perfect, pluperfect) with precision
- Shift between indicative and subjunctive to convey stance, certainty, or nuance
- Use complex temporal relationships and aspectual periphrases
- Employ sophisticated discourse markers for cohesion
- Create long, multi-clause sentences with embedded structures
- Use idiomatic expressions, collocations, and register-appropriate language
- Handle hypothetical past situations (Si hubiera sabido, habría venido)
""",

        "C2": """
**LLM Output Guidelines:**
- Demonstrate complete command of all grammatical structures
- Use grammar choices for stylistic effect and subtle meaning distinctions
- Employ nominalizations and reduced clauses for elegant expression
- Navigate formal/informal registers seamlessly
- Use advanced passive constructions and "se" variations appropriately
- Include specialized vocabulary and complex syntax naturally
- Express ideas with precision, fluidity, and near-native nuance
""",

        "Native": """
**Native Speaker Mode - No Constraints:**
Speak naturally as a native Spanish speaker would in casual conversation. No restrictions on grammar, vocabulary, complexity, or style. Express yourself freely and authentically without simplification or pedagogical modifications.
"""
    ]
}
